### DISCUSSION TITLE
Отображение текста без прокрутки

### DISCUSSION THREAD

**Discussion by vmashchenko** [vvip-68/LedPanelWiFi#2] [Общие вопросы]
Приветствую!

Спасибо за подробные [инструкции](https://github.com/vvip-68/LedPanelWiFi/wiki/%D0%9D%D0%B0%D1%81%D1%82%D1%80%D0%BE%D0%B9%D0%BA%D0%B0-%D0%B1%D0%B5%D0%B3%D1%83%D1%89%D0%B5%D0%B9-%D1%81%D1%82%D1%80%D0%BE%D0%BA%D0%B8). Не нашёл только описания того, можно ли выводить просто статический текст на экран, чтобы он стоял на месте, и никуда не бежал? И как это сделать?

Может это можно сделать с помощью макроса {Ts}, указав время показа строки, например 1.000.000 секунд?

Второй вариант: текст выезжает на экран и остается там стоять (определенное количество времени, или пока не поступит иная команда по UDP). Такого себе рода информационное табло.

Вячеслав

**Comment by vvip-68** [comment]
Нет такого функционала.
А то делать, когда текст длиннее экрана?

**Reply by vmashchenko** [reply]
Обрезать.

Отображение статического текста - это стандартный функционал для LED-панелей/табло - на улицах повсеместно используется.
Хорошо бы иметь в прошивке.

Обычно используется в комбинации - текст выезжает и останавливается. Или один блок текста маленький. всегда отражается (температура / время), а основной блок - бежит строка.

**Reply by vvip-68** [reply]
Это не панель для текстов. Это - новогодняя гирлянда на окно...

**Reply by vmashchenko** [reply]
Ок. А почему бы не задержать, например, какое-то важное сообщение на какое-то время?
Поздравление с Новым годом показать, оставить на 10-20 секунд, помигать разным цветом? Поанимировать. Будет ещё круче.

Или часы отображать регулярно. Или тот же обратный отсчёт времени - держать на одном месте.

**Reply by vvip-68** [reply]
Потому что все тот же вопрос не решенный для меня - что делать с текстом, который в 99% случаев шире чем видимая область  матрицы. Часы регулярно - либо эффект как таковой "Часы", либо включенная галочка "Часы поверх эффектов".  Опять же - 95% известных мне устройств - это именно часы на стандартной матрице 16x16, либо на сдвоенной такой матрице (не считая гирлянд, который как правило делаются на китайских "нитках" по 50 диодов с разрешением гирлянды примерно шириной 20-25 и высотой 10 диодов. На ширине 20-25 можно одновременно показать 4 буквы. Так себе текст... Обратный отсчет времени - очень специфичный и мало кому нужный "эффект", с ним проще - но не представляю зачем он нужен.

Тексты с эффектами , выплыванием, анимацией - делаются как ролики в Jinx, сохраняются на SD-карту и потом "крутятся" на панели в эффекте SD-карта.

**Reply by vvip-68** [reply]
Адресная лента (а это именно проект панелей на адресной ленте) имеет ограничения по времени вывода информации на всю длину цепочки диодов. На длине более 2000 диодов для наиболее типичной WS2812 уже будет заметное торможение в выводе - низкий FPS. Для текстовых панелей, чтобы текст хорошо читался, нужна плотная упаковка диодов типа как на промышленных панелях P10. На них текст будет хорошо читаться, можно каскадировать, достаточное быстродействие - даже экраны делают. Но это совсем другая технология и даже на других микроконтроллерах. К примеру ESP32 насколько я просматривал этот вопрос - едва ли потянет такую панель более чем на 128x64, максимум 128x128... Нужен другой специализированный контроллер, а не вот это вот все DIY

**Reply by vmashchenko** [reply]
Не думали прошивку адаптировать и под матрицы P10, о которых пишете? Это было бы совсем круто.
Мне попадалась прошивка эквалайзера (достаточно простая) на esp8266, которая могла работать и на ws2812 и на p10 подобных матрицах. Могу показать.

Что касается текста, у меня сейчас матица на 2700 пикселей 112х24. Хорошо там работают анимации, бегущая строка и веб-панель. Это очень красиво.
На esp32.
Скорость, конечно, ниже, чем на матрицах 16*16, но работает! И эффект огня достойно смотрится. 
Я ее проектировал с учётом вывода 10-11 символов двойной толщины!
Но вот не могу найти хорошую прошивку, чтобы она отображала и статический текст, помимо бегущего и анимаций.

По поводу скорости, - я немного почитал, - обычно на больших матрицах делают параллельный вывод (не один дата пин, а 2-3-4…). Библиотека FastLed(Neopixel) это поддерживает, просто добавляется ещё одна строка для каждого дата пина. Больше программного кода не требуется.

Теоретически, если по 2 тыс. светодиодов вешать на один пин, то можно и 10 тыс. светодиодов сделать.

**Reply by vvip-68** [reply]
Вывод на матрицы P10 - совсем другая технология, совсем другие библиотеки. И это не FastLED. А здесь функции FastLED используются во многих эффектах, в частности в шумовых именно для построения изображения.  Ссылку дайте, да. Будет время - посмотрю как там и что.

Написание какого-то проекта (или адаптация) - как правило преследуют какую-то цель. Вроде - "Мне нужна гирлянда на окно на новый год. Почему бы и не сделать? О! Это можно и как часы использовать". Или вот - интересно поковыряться как делается то-то и то-то, поковыряем... Или - есть заказчик, хорошие деньги. По срокам не жестит, почему бы не заняться...

А вот адаптация / написание для P10... Ну, разве что из интереса. А на практике - табло и бегущие строки мне не нужны. Разве что очередные часы повторить (даже не сделать, а из готовых проектов типа MarioClock или TetrisClock). Ну ок, заказал панельки 64х64 и 32х64 для часов. К концу июля алиэкспресс обещал доставить. А для проектов табло длинных строк, хотя бы для тестирования нужна не одна такая матрица, а штук 10, что не дешево. И не факт, что ESP32 потянет такое количество матриц. Мне не встречались проекты на ESP32 больше 128 точек в ширину...

По поводу скорости - и текущая версия проекта поддерживает параллельный вывод. Я сделал на 4 линии, конфигурируемые из Web. Впрочем, несложно расширить и до 8... Или просто в коде эти 8 строк инициализации прописать. Только теория не всегда хорошо коррелирует с практикой. Разделение на линии может и не дать эффекта. Все-таки ESP32 довольно слабенький для формирования/ обработки большого видео-массива.

112 в ширину...  10-11 символов двойной толщины. Какой, к примеру, содержательный статический текст из 11 символов можно отображать?

**Reply by vmashchenko** [reply]
Ссылка: https://github.com/donnersm/FFT_ESP32_Analyzer
Видео: https://www.youtube.com/watch?v=bQ7c9Vlhyp0

> А вот адаптация / написание для P10... Ну, разве что из интереса. А на практике - табло и бегущие строки мне не нужны.

Возможно, будет интересно с точки зрения плотности пикселей и их количества. Эффекты круче будут смотреться!

> 112 в ширину... 10-11 символов двойной толщины. Какой, к примеру, содержательный статический текст из 11 символов можно отображать?

Почти помещается "С Новым Годом!")). Ну если в один пиксел делать ширину букв, то влезет уже 20 символов. У меня стояла задача сделать матрицу на 10 пикселей, которые видны издалека.

**Reply by vmashchenko** [reply]
Несколько простых приземленных вопроса тогда к Вам:

1. Где/как в коде можно остановить строку, чтобы она просто перестала бежать?
2. Как можно узнать текущую позицию первого пиксела строки?
3. Как можно рассчитать длину строки в пикселах, чтобы знать в каком месте остановить строку, чтобы она оказалась по центру?

Попробую сам это сделать в Вашем коде, тем более на него приятно смотреть) Много комментариев и хорошо читается!

И вопрос по макросам:
4. {Ts} - что означает время показа строки?
5. Можно ли менять скорость движения бегущей строки на ходу?
Если да, то, теоретически, можно было бы в какой-то момент, когда часть строки выехала просто ее остановить, поставив скорость 0.

**Reply by vvip-68** [reply]
Переменная offset, объявлена в a_soft_def.h
В начале инициализации строки (начало показа)  - равна ширине матрицы - выезжает с правого края, далее текст сдвигается на 1 по таймеру - по сути - каждый вызов отрисовка строки - fillString() в runningText.ino переменная offset уменьшается на 1.
Известна ширина одного символа строки 5, 8 или 10 в зависимости от выбранного типа/размера шрифта, известен пробел между символами - 1 (для 5,8) или 2 (для 10) точки. Очищаете строку от макросов (кажется currentText - уже очищенная), домножаете кол-во символов в строке на ширину символа с учетом расстояния между буквами. Ширина матрицы - известна. Дальше считайте.

{Ts} - ну вот то и означает сколько секунд будет отображаться строка на экране, пока не уберется. Прокрутили, строка закончилась, время не вышло? Крутим снова.

Таймер - из библиотеки Гайвера про таймеры - называется timerText. Библиотекa GyverTimer папке с библиотеками. Интервал срабатывания - вызов  textTimer.setInterval(textScrollSpeed) - в миллисекундах.

Остановить можно, видимо, либо установить таймер в большое значение (не в 0 - 0 это 0мс - будет срабатывать на каждом цикле), либо перестав уменьшать значение offset.

Сделаете - прошу выложить видео, как это будет смотреться.

**Reply by vmashchenko** [reply]
Спасибо. Хорошо.

Так сейчас выглядят эффекты на большом экране, если интересно:
https://youtu.be/j-lSnAGHKgM
https://youtu.be/jYeA1-tAGMI
https://youtu.be/nhdxDbUswxg

**Reply by vvip-68** [reply]
Ну, неплохо. Но у меня таких эффектов нет, значит это либо чья-то другая прошивка (чья? - м.б. утянуть эффекты) - либо это ролики на SD-карте).

**Reply by vmashchenko** [reply]
Эта прошивка: https://community.alexgyver.ru/threads/wifi-lampa-budilnik-obsuzhdenie-proshivki-ot-gunner47.2418/
В коде той прошивки была указаны ссылка на эту прошивку (на esp8266), так я её и нашел))

Это именно эффекты. Без SD карты.

**Reply by bestsale** [reply]
Ребят приветсвую можно ваш контакт как связаться? Хотел бы попросить с проектом помочь с прошивкой, не бесплатно конечно 
Спасибо
